### PR TITLE
[RAPTOR-5841][RAPTOR-5842][RAPTOR-6190][RAPTOR-6230] R: Fix sparse data support, fix sampling issue

### PR DESCRIPTION
**Pull request by jmbannon**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Sparse support in R never really existed - it would ignore the colnames and always cast the data to dense. This PR ensures the matrix stays sparse, and passes the column names file to it.

In addition, an R bug is fixed which sampled X and y differently (every R model up to this point has been worthless.. lol)

## Changes
5841,5842:
- Do not cast readMM to dataframe, leave it as dgTMatrix
- Actually pass sparse input matrix colnames
- Allow transforms to output a sparse matrix
  - If the output is sparse, convert it to a coordinate-format dataframe (py2r has no r-matrix-to-py support, only df afaik)
- Create R sparse regression template, use it in tests

6190:
- Only call sample once, use it on all input data
- Drop weight rows where y value is NA

### CODE REVIEW

**Review by scottp-ml — CHANGES_REQUESTED** [review]
Can you please add a model_metadata with typeschema for the new example?

File: custom_model_runner/datarobot_drum/resource/components/Python/r_fit/fit.R [hunk 1]
```diff
@@ -40,57 +40,77 @@ init <- function(code_dir) {
...
+        }
 
+        # Then, drop all rows where the y value is NA
+        na_rows <- as.vector(is.na(y[target_name]))
+        X <- X[!na_rows,, drop=FALSE]
+        y <- y[!na_rows,, drop=FALSE]
+    # If no target is provided, then treat the df as X
     } else {
+        X <- df
+        na_rows <- rep(FALSE, nrow(X))  # Set all of na_rows to FALSE since no y values are NA
```
Comment (jmbannon) [inline custom_model_runner/datarobot_drum/resource/components/Python/r_fit/fit.R:94]: Add print if NA rows are dropped

File: tests/fixtures/r_estimator_sparse.R [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  }
+
+  # Save model
+  model_path <- file.path(output_dir, 'artifact.rds')
+  saveRDS(c("dummy", "artifact"), file = model_path)
+}
+
+score <- function(X, model, ...) {
+   # return a data.frame with a 'Predictions' column with a single value of 42
+   data.frame('Predictions'=rep(42, nrow(X)))
```
Comment (jmbannon) [inline tests/fixtures/r_estimator_sparse.R:16]: add spare + colname check in score

File: tests/drum/test_fit.py [hunk 3]
```diff
@@ -556,16 +557,18 @@ def test_fit_simple(
         )
 
     @pytest.mark.parametrize(
-        "framework", [SKLEARN_SPARSE, PYTORCH, RDS,],
+        "framework, problem, language, is_training",
+        [
+            (SKLEARN_SPARSE, SPARSE, PYTHON, True),
+            (PYTORCH, SPARSE, PYTHON, True),
+            # This is testing an R custom task, but we have to set is_training to False because that's how you convert
```
Comment (4dahalibut) [inline tests/drum/test_fit.py:564]: cursed

**Review by 4dahalibut — COMMENTED** [review]
please add tests into predict hook for R test.

File: tests/fixtures/r_estimator_sparse.R (hunk 2 above)
Comment (jmbannon) [reply tests/fixtures/r_estimator_sparse.R]: we don't actually pass sparse colnames to the score hook (including python).
Made a ticket: https://datarobot.atlassian.net/browse/RAPTOR-6231

**Comment by jmbannon** [comment]
@4dahalibut @scottp-ml Added your comments. Ran into more issues along the way so please review the latest commit

**Review by scottp-ml — APPROVED** [review]

**Comment by jmbannon** [comment]
jarvis test functional_inference_model_templates functional_training_model_templates

**Comment by jmbannon** [comment]
jarvis test functional_inference_model_templates functional_training_model_templates

**Pull request merged**